src/System/src/FloatingBaseSystemDynamics.cpp

```diff
@@ -74,8 +74,25 @@ bool FloatingBaseDynamicalSystem::setMassMatrixRegularization(const iDynTree::Ma
     return true;
 }
 
-bool FloatingBaseDynamicalSystem::dynamics(const StateType& state,
-                                           const double& time,
+bool FloatingBaseDynamicalSystem::setState(const StateType& state)
+{
+
+    std::get<0>(m_state) = std::get<0>(state);
+    std::get<1>(m_state) = std::get<1>(state);
+    std::get<2>(m_state) = std::get<2>(state);
+    std::get<4>(m_state) = std::get<4>(state);
+
+    // project the base orientation matrix in SO3
+    // here we assume that the velocity is expressed using the mixed representation
+    const Eigen::Matrix3d& baseOrientation = std::get<3>(state);
+    Eigen::JacobiSVD<Eigen::Matrix3d> svd(baseOrientation,
+                                          Eigen::ComputeFullU | Eigen::ComputeFullV);
+    std::get<3>(m_state) = svd.matrixU() * svd.matrixV().transpose();
```

Comment:
This can be slow, just to understand  did you check if that is not a bit bottleneck?

Reply:
Well @traversaro you are actually right. Half of the time is spent in the `setState()` function and the bottleneck is the `svd` calculation.

![calgrind](https://user-images.githubusercontent.com/16744101/87177805-86e57600-c2dc-11ea-9c1e-e7fddd05abf0.png)

Do you think that a [Baumgarte stabilization](https://ieeexplore.ieee.org/document/7402298?reload=true) may solve the problem?

Reply:
I don't know what the best option is at the moment, but if this PR is just about using Eigen perhaps we can leave this normalization stuff in a following PR? Or just leave the JacobiSVD  for now. 

BTW I just found https://github.com/ericjang/svd3, probably not the best option for maintainability but quite a neat idea.

Reply:
Do you want to perform the integration on SO(3)? Why not passing through quaternions?

Reply:
For sure quaternion are much easier to normalize.

Reply:
Also, why you are performing this kind of computations in ``setState``? Is it to make sure that the user is actually passing a rotation matrix?

Have you considered implementing the computations for SVD manually, since the matrix is small?

Reply:
@S-Dafarra 
> Do you want to perform the integration on SO(3)? Why not passing through quaternions?
working on this

> Also, why you are performing this kind of computations in setState? Is it to make sure that the user is actually passing a rotation matrix?
This is the `setState()` of the `FloatingBaseSystemDynnamics` class so the user has to pass a rotation

Reply:
Actually I can remove the svd If we switch to quaternions and I use this equation to compute the quaternion derivative
![image](https://user-images.githubusercontent.com/16744101/87328391-b347ff00-c535-11ea-8087-ad4a487f9139.png)

Reply:
Actually using quaternion is not so trivial.

Indeed If I use `Eigen::Quaterniond` I can't do this:
```
Eigen::Quaterniond a;
a.setIdentity();
a = a + Eigen::Vector4d::Zero();
``` 
I get the following 
```
 error: no match for ‘operator+’ (operand types are ‘Eigen::Quaterniond {aka Eigen::Quaternion<double>}’ and ‘const ConstantReturnType {aka const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, 4, 1> >}’
```

Unfortunately, the `addArea()` is a template function and I would avoid adding template specialization there.
For the time being, I would go for the Rotation matrix + Baumgard integration.  In the future, If we want to use quaternions instead of matrices we can open a new issue and discuss it there. What do you think?

Reply:
I am a little lost here. The state derivative and the state are generally supposed to be different right? In this case the state is a quaternion and the derivative a Vector4 and it makes sense for them not to be added together. Before you were dealing with the rotation matrix in a specific case. How would it work now?

Reply:
Let me explain better. The dynamical system class contains the state, its derivative, and the input. In case of a rotation we have:
- `Eigen::Matrix3d` for the state (i.e. the rotation) 
- `Eigen::Matrix3d` for the state derivative (i.e. the rotation rate of change) 
- `Eigen::Vector3d` for the input (i.e. angular velocity)

The rate of change the rotation matrix is then computed as
<p align="center">
Ṙ = ωx R 
</p>
Ṙ is then integrated to compute R. The `ForwardEuler` integrator uses the function `addArea()` to compute the integral 
https://github.com/dic-iit/bipedal-locomotion-framework/blob/339cf76339622cf496d941258374614f43d526af/src/System/include/BipedalLocomotion/System/ForwardEuler.h#L48
This code can compile only if the `operator+()` is defined between the state and the state derivative. In this case, since the rate of change of the rotation matrix and the rotation matrix are stored in an `Eigen::Matrix3d`, the `operator+()` is well defined. 

In the case of quaternions we can decide to:
1. Store the quaternion in an `Eigen::Vector4d` and apply a similar approach. In this case, I won't have any problems;
2. Store the quaternion in an `Eigen::Quaterniond` and its rate of change in `Eigen::Vector4d`. In this case `std::get<I>(x) += std::get<I>(dx) * dT` will fail because the `operator+()` is not defined between `Eigen::quaterniond` and `Eigen::Vector4d`. 


By the way, in https://github.com/dic-iit/bipedal-locomotion-framework/pull/70/commits/9870a4f218f4f42938c66d29cb6b4bebe2a9ffa2 I implemented the Baumgarte stabilization for rotation matrix. Differently from the [original paper](https://ieeexplore.ieee.org/document/7402298?reload=true), I derived the equation in case of angular velocity expressed in the inertial frame. Thanks to this commit I was able to remove the `svd` computation from the `setState()` and consequentially I improved the overall performances of the classes.

Reply:
Great. This has also the nice property that the simulated  system remains an ODE, while with the normalization you are basically coupling an ODE with an discrete non linear instantaneous system, i.e. the block that takes in input the integrated orientation (that is  the output of the ODE) and  returns the normalized rotation instead.

Reply:
That's very clear, thank you! In the hindsight, it is also pretty clean to have directly the rotation matrix derivative, nice solution.